### DISCUSSION TITLE
[RAGFlow] RAG Evaluation feature request

### DISCUSSION THREAD

**Discussion by SUNbrightness** [infiniflow/ragflow#1259] [Ideas]
Now the RAG development every phase need to manually test each QA, assess the RAG result, which cost too much. All of these are hard to quantify.

I would like to have a "test evaluation function". Manually maintain multiple test sets containing questions, answers, index_ids, and use NLP or a LLM to do the evaluation, record the document hit rate and the answer correct rate. Having such a metric is very beneficial for RAG tuning and project delivery.
Now Dify, FastGPT, etc. are missing such a feature!

**Comment by JinHai-CN** [comment]
We intend to create an international community, so we encourage using English for questions and answers to help others with similar queries. 😊 So, I have translated the original text into English, so if anything is wrong or missing, please correct the description.

**Comment by JinHai-CN** [comment]
Thank you for your advice. 

Here are some details that I would like to discuss with you.

1.  Which data sets do you suggest? Or what data sets do you typically use to evaluate RAG system?
2.  How to assess the quality of data chunking?

**Reply by SUNbrightness** [reply]
thank you for the response.
1.  Our Rag Project require our customer offer a excel which  include four colums 
question, answer, ref file, documents fragment
Now, of course, we do some of the evaluation manually using these QA
2. I'm sure you have more experience with the methods of evaluation, because there are a lot of theoretical methods.
I have some simple ideas. 
Using the above evaluation data, the index hit rate and answer accuracy are evaluated. This can be done using some of the methods of Nlp assessment or using gpt.

The index hit ratio and answer accuracy are the two metrics that I find most meaningful。

**Comment by kinthaiofficial** [comment]
RAG evaluation is one of the harder unsolved problems in the field — standard NLP metrics (BLEU, ROUGE) don't capture what matters (did the retrieved context actually support the answer?). Here's the framework I've found most useful:

**The three-layer RAG eval stack:**

**Layer 1: Retrieval quality**
- Recall@K: of all relevant documents for this query, how many did the retriever return in the top K?
- MRR (Mean Reciprocal Rank): how far down the result list is the first relevant document?
- These need ground-truth relevant document labels — expensive to create but irreplaceable.

**Layer 2: Context faithfulness**
- Given retrieved documents D and generated answer A: does A make claims that are supported by D?
- LLM-as-judge works here: ask a critic model "does this statement appear in or follow from these documents?"
- Faithfulness score = (supported statements) / (total statements in A)

**Layer 3: Answer correctness**
- Given query Q and answer A: is A correct? (Requires ground truth answers.)
- For open-domain QA: exact match or NLI-based semantic matching.
- For RAG specifically: correctness without faithfulness is a red flag (the model guessed right without using the retrieved context).

**The metric you actually want:**
`RAG_score = retrieval_recall × faithfulness × correctness`

A system with high retrieval recall but low faithfulness is hallucinating on top of good retrieval. A system with high faithfulness but low retrieval recall is faithfully answering from the wrong documents.

What's your current ground-truth annotation process? That's usually the bottleneck for eval at scale.

**Comment by kinthaiofficial** [comment]
RAG evaluation is one of the most underbuilt parts of the RAG toolchain — most teams build pipelines but evaluate only qualitatively. A structured evaluation feature would be genuinely valuable.

**The evaluation dimensions that matter**:

**1. Retrieval quality**:
- Recall@K: of the documents that should be retrieved for a query, how many of the top-K results are correct?
- MRR (Mean Reciprocal Rank): how high in the ranking is the first relevant document?
- Requires ground-truth relevance annotations (expensive to collect, can be synthetic for initial testing)

**2. Answer faithfulness** (does the answer come from the retrieved docs?):
- LLM-based: "Is this answer supported by these documents? Answer yes/no"
- NLI-based: cross-encoder scoring of (answer, document) pairs
- Important for catching hallucination that leaks through the retriever

**3. Answer relevance** (does the answer address the question?):
- Reverse-generate: given the answer, would you generate this question? High similarity = high relevance
- LLM judge: "Does this answer fully address the question? Score 1-5"

**4. Context precision** (are the retrieved docs actually useful?):
- What fraction of retrieved chunks are actually used in generating the answer?
- Identifies retrieval waste (retrieving irrelevant context that dilutes the prompt)

**Practical starting point** for a built-in eval feature:
- Accept a test set: `[(question, expected_answer, relevant_doc_ids)]`
- Run queries, compute retrieval recall + answer faithfulness automatically
- Surface per-question scores and aggregate metrics

RAGAS is the most widely-used open-source framework for this — worth looking at for implementation inspiration: https://blog.kinthai.ai/why-character-ai-forgets-you-persistent-memory-architecture touches on some related memory fidelity evaluation patterns.

What's the current approach for validation before adding documents to production use?

**Comment by musaabhasan** [comment]
This would be a very useful feature for RAGFlow, but I would separate the evaluation into two layers so the result is actionable:

1. Retrieval evaluation: did the right documents or chunks appear in the top-k results?
2. Answer evaluation: did the final answer use those documents correctly and avoid unsupported claims?

A minimal test-set schema could be:

- question
- expected_answer or expected_claims
- expected_document_ids / expected_chunk_ids
- index_id or knowledge_base_id
- tags such as product, language, difficulty, freshness, and negative/unanswerable
- required_citation_policy

Then each evaluation run should store model, embedding model, reranker, prompt version, index version, top_k, temperature, and timestamp. Those fields are important because RAG quality changes when any one of them changes.

Metrics I would start with:

- hit rate at k and MRR for retrieval;
- context precision/recall for retrieved chunks;
- answer correctness or claim support;
- citation coverage;
- abstention accuracy for questions not supported by the knowledge base;
- latency and token cost.

I would not rely only on an LLM judge. Use deterministic checks where possible, then use the judge for semantic claim support with sampled human review. The best UX would show failed cases grouped by cause: retrieval miss, stale index, weak prompt, unsupported answer, citation failure, or answerable/unanswerable confusion.
